lib/bap_disasm/bap_disasm_reconstructor.ml

```diff
@@ -19,51 +19,45 @@ type reconstructor = t
 let create f = Reconstructor f
 let run (Reconstructor f) = f
 
-let find_calls name roots cfg =
-  let starts = Addr.Table.create () in
-  List.iter roots ~f:(fun addr ->
-      Hashtbl.set starts ~key:addr ~data:(name addr));
-  Cfg.nodes cfg |> Seq.iter ~f:(fun blk ->
-      let () =
-        if Seq.is_empty (Cfg.Node.inputs blk cfg) then
-          let addr = Block.addr blk in
-          Hashtbl.set starts ~key:addr ~data:(name addr) in
-      let term = Block.terminator blk in
-      if Insn.(is call) term then
-        Seq.iter (Cfg.Node.outputs blk cfg)
-          ~f:(fun e ->
-              if Cfg.Edge.label e <> `Fall then
-                let w = Block.addr (Cfg.Edge.dst e) in
-                Hashtbl.set starts ~key:w ~data:(name w)));
-  starts
+let callees_of_block cfg roots blk =
+  let addr = Block.addr blk in
+  let term = Block.terminator blk in
+  let init =
+    if Set.mem roots addr then Block.Set.singleton blk
+    else Block.Set.empty in
+  if Insn.(is call) term then
+    Seq.fold ~init (Cfg.Node.outputs blk cfg)
+      ~f:(fun cls e ->
+          if Cfg.Edge.label e <> `Fall then
+            Set.add cls (Cfg.Edge.dst e)
+          else cls)
+  else init
+
+let update_callees cfg roots callees blk =
+  Set.union callees (callees_of_block cfg roots blk)
+
+let find_callees cfg roots =
+  let roots = Addr.Set.of_list roots in
+  Seq.fold (Cfg.nodes cfg) ~init:Block.Set.empty
+    ~f:(update_callees cfg roots)
 
 let reconstruct name roots cfg =
-  let roots = find_calls name roots cfg in
-  let init =
-    Cfg.nodes cfg |> Seq.fold ~init:Cfg.empty ~f:(fun cfg n ->
-        Cfg.Node.insert n cfg) in
-  let filtered =
-    Cfg.edges cfg |> Seq.fold ~init ~f:(fun cfg e ->
-        if Hashtbl.mem roots (Block.addr (Cfg.Edge.dst e)) then cfg
-        else Cfg.Edge.insert e cfg) in
-  let find_block addr =
-    Cfg.nodes cfg |> Seq.find ~f:(fun blk ->
-        Addr.equal addr (Block.addr blk)) in
-  Hashtbl.fold roots ~init:Symtab.empty
-    ~f:(fun ~key:entry ~data:name syms ->
-        match find_block entry with
-        | None -> syms
-        | Some entry ->
-          let cfg : cfg =
-            with_return (fun {return} ->
-                Graphlib.depth_first_search (module Cfg)
-                  filtered ~start:entry ~init:Cfg.empty
-                  ~enter_edge:(fun _ -> Cfg.Edge.insert)
-                  ~start_tree:(fun n t ->
-                      if Block.equal n entry
-                      then Cfg.Node.insert n t
-                      else return t)) in
-          Symtab.add_symbol syms (name,entry,cfg))
+  let callees = find_callees cfg roots in
+  let is_call e = Set.mem callees (Cfg.Edge.dst e) in
+  let rec traverse fng node =
+    let fng = Cfg.Node.insert node fng in
```

Comment:
Please, don't use abbreviations that are not common, suggestions,

`graph` or `g` or `prog` for the whole program graph
`cfg` for the CFG. 

Also, not `edg`, either `edge` or `e`.